internal/overlord/servstate/manager.go

```diff
@@ -185,22 +186,62 @@ func (m *ServiceManager) Ensure() error {
 	return nil
 }
 
-// ActiveServices returns the name of the services which are currently
-// set to run. They may be running or not depending on the state of their
-// process lifecycle.
-func (m *ServiceManager) ActiveServices() []string {
-	var names []string
+type ServiceInfo struct {
+	Name    string
+	Default plan.ServiceAction
+	Status  ServiceStatus
+	Message string
+}
+
+type ServiceStatus string
+
+const (
+	ServiceActive   ServiceStatus = "active"
+	ServiceInactive ServiceStatus = "inactive"
+	ServiceError    ServiceStatus = "error"
+)
+
+// Services returns the list of configured services and their status, sorted
+// by service name. Filter by the specified service names if provided.
+func (m *ServiceManager) Services(names []string) ([]*ServiceInfo, error) {
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return nil, err
+	}
+	defer releasePlan()
+
+	requested := make(map[string]bool)
+	for _, name := range names {
+		requested[name] = true
+	}
+
+	var services []*ServiceInfo
 	for name, service := range m.plan.Services {
+		if len(names) > 0 && !requested[name] {
```

Comment:
This is convoluted logic to express pretty important functionality for this API method.

I'd prefer:

```
onlyMatchRequested := len(names) > 0
for name, service := range m.plan.Services {
    if onlyMatchRequested && !requested[name] {
        continue
    }
```

Reply:
Agreed, except I'd invert the logic and say simply "all := true/false" to make it more obvious.

Reply:
"all" actually led to more complex/inverted boolean logic (`!all && !requested[name]` or `!(all || requested[name])`, so I kept the logic the same but called it `matchNames` so it reads better but isn't too long.